lib/plugins/dartdoc/plugin.dart

```diff
@@ -0,0 +1,346 @@
+      for (final map in content.cast<Map<String, dynamic>>())
+        if (!map.containsKey('__PACKAGE_ORDER__')) DartdocEntry.fromJson(map),
+    ];
+
+    _documentationCache[package] = (now, entries);
+    return (urlBase, entries);
+  }
+
+  final _searchCache = <String, (DateTime, List<String>)>{};
+
+  /// Search for packages based on [query] on https://pub.dev.
+  Future<List<String>> searchPackages(String query) async {
+    // Most common case.
+    if (query == 'flutter') return ['flutter'];
+
+    // A cache isn't really needed here (since the most common case by far is
+    // hard coded above) but it avoids triggering multiple requests when a
+    // package's documentation is referenced multiple times in one message.
+    final cached = _searchCache[query];
+    final now = DateTime.timestamp();
+    final cacheExpiry = now.add(-expireDuration);
+
+    if (cached != null && cached.$1.isAfter(cacheExpiry)) {
+      return cached.$2;
+    }
+
+    final response = await http.get(
+      Uri.https('pub.dev', '/api/search', {'q': query}),
+    );
+    final content = jsonDecode(utf8.decode(response.bodyBytes)) as Map;
+
+    final results = (content['packages'] as List)
+        .cast<Map<String, Object?>>()
+        .map((e) => e['package'] as String)
+        .toList();
+
+    _searchCache[query] = (now, results);
+    return results;
+  }
+
+  @override
+  void afterConnect(NyxxGateway client) {
+    final parser = SearchGrammar().build<List<Search>>();
+
+    Future<void> handle(Search search, MessageCreateEvent event) async {
+      switch (search.kind) {
+        case SearchKind.elementLookup:
+        case SearchKind.elementSearch:
+          final matchingPackages = await searchPackages(search.package!);
+
+          if (matchingPackages.isEmpty) {
+            await event.message.channel.sendMessage(MessageBuilder(
+              replyId: event.message.id,
+              allowedMentions: AllowedMentions(repliedUser: false),
+              embeds: [
+                EmbedBuilder(
+                  color: const DiscordColor.fromRgb(255, 0, 0),
+                  title: 'Package Not Found',
+                  description: search.package!,
+                ),
+              ],
+            ));
+            break;
+          }
+
+          final package = matchingPackages.first;
+
+          final (urlBase, entries) = await getEntries(package);
+
+          final useNamePattern = RegExp(r'^([A-Za-z$_A-Za-z0-9$_]*?)$');
+          final query = search.name;
+          final byName = useNamePattern.hasMatch(search.name);
+
+          // lehvenstein can block, especially if we are searching the Flutter
+          // docs (80k+ calls). We need to move the isolate creation to a
+          // different function to avoid capturing `this` in the closure
+          // context.
+          Future<List<DartdocEntry>> process(
+            String query,
+            List<DartdocEntry> entries, {
+            required bool byName,
```

Comment:
What do you think about pulling this out as its own private top-level function? I think it would make this method much easier to visually parse.